File: core/core/controller.py
import asyncio
import time

from core.logger import custom_logger


# проверить логи, возможно предел будет пробиватся, если нет то повысить до 6000

class BinanceAPIController:
    """
    данный контроллер нужен чтобы не пробить весовой лимит BinanceApi и не получить автобан
    # позже добавить учет вебсокет соединений | лимит - 300 подключений за попытку каждые 5 минут
    """
    def __init__(self, max_weight: int = 5700):
        """возможно слишком трудно будет это развивать дальше"""
        self.max_weight = max_weight # на самом деле лимит 6000, но лучше сделать поправку на асинхронность
        self.current_weight = 0
        self.last_reset_time = time.time()
        self.lock = asyncio.Lock()
        self.queue = asyncio.Queue()
        self.lock = asyncio.Lock()
        self.queue = asyncio.Queue()

    async def start(self):
        asyncio.create_task(self.reset_loop())
        asyncio.create_task(self.process_queue())

    async def reset_loop(self): # автосброс(скорее всего слишком затратный)
        while True:
            await asyncio.sleep(60)
            async with self.lock:
                self.current_weight = 0
                self.last_reset_time = time.time()

    async def process_queue(self):
        """Обрабатывает очередь запросов (если лимит был превышен)"""
        while True:
            request = await self.queue.get()
            await request()
            self.queue.task_done()

    async def request_with_limit(self, weight: int, request_func):
        """Управляет лимитом и выполняет запрос. Если лимит превышен, запрос ставится в очередь."""
        async with self.lock:
            if self.current_weight + weight > self.max_weight:
                custom_logger.log_with_path(
                    level=1,
                    msg=f"reached the limit for Binance API. Current weight:  {self.current_weight}",
                    path="ApiLimits.log"
                )
                await self.queue.put(lambda: request_func())
                return None

            self.current_weight += weight

        return await request_func()

controller = BinanceAPIController()

File: crypto-gateway/internal/modules/strategy/service/controller.go
package service

import (
	"crypto-gateway/internal/appmetrics"
	"fmt"
	"log"
	"sync"
	"time"
)

type RequestFunc func() error

// данный контроллер следит за тем чтобы не пробить весовой лимит Binance.com и не получить автобан.
// Также он отвечает за выполнение запросов из структуры BinanceApi
type BinanceAPIController struct {
	MaxWeight       int
	CurrentWeight   int
	lastResetTime   time.Time
	Mu              sync.Mutex
	queue           chan queuedRequest
	queueEvent      chan struct{}
	pendingRequests map[string]struct{}
}

type queuedRequest struct {
	id     string
	fn     RequestFunc
	weight int
}

func NewBinanceAPIController(maxWeight int) *BinanceAPIController {
	c := &BinanceAPIController{
		MaxWeight:       maxWeight,
		CurrentWeight:   0,
		lastResetTime:   time.Now(),
		queue:           make(chan queuedRequest, 1000), // configurable buffer
		queueEvent:      make(chan struct{}, 1),
		pendingRequests: make(map[string]struct{}),
	}

	go c.resetLoop()
	go c.processQueue()
	return c
}

// автосброс нагрузки на бинанс
func (c *BinanceAPIController) resetLoop() {
	ticker := time.NewTicker(60 * time.Second)
	defer ticker.Stop()
	for {
		<-ticker.C
		c.Mu.Lock()
		c.CurrentWeight = 0
		c.lastResetTime = time.Now()
		c.Mu.Unlock()
	}
}

// обрабатывает очередь запросов
func (c *BinanceAPIController) processQueue() {
	for range c.queueEvent {
		for {
			c.Mu.Lock()
			if len(c.queue) == 0 {
				c.Mu.Unlock()
				break
			}
			req := <-c.queue
			delete(c.pendingRequests, req.id)
			c.Mu.Unlock()

			if err := req.fn(); err != nil {
				appmetrics.AnalyticsServiceLogging(2, "Error executing queued request", err)
				log.Printf("Error executing queued request: %v", err)
			}

			c.Mu.Lock()
			c.CurrentWeight += req.weight
			c.Mu.Unlock()
		}
	}
}

// Управляет лимитом и выполняет запрос. Если лимит превышен, запрос ставится в очередь.
func (c *BinanceAPIController) RequestWithLimit(weight int, fn RequestFunc) error {
	c.Mu.Lock()
	if c.CurrentWeight+weight > c.MaxWeight { // PUSH уведомление добавить
		appmetrics.AnalyticsServiceLogging(2, fmt.Sprintf("API limit reached. Current weight: %d", c.CurrentWeight), nil)
		log.Printf("API limit reached. Current weight: %d", c.CurrentWeight)

		req := queuedRequest{fn: fn, weight: weight}
		c.queue <- req

		select {
		case c.queueEvent <- struct{}{}:
		default:
			// Already triggered
		}

		c.Mu.Unlock()
		return nil
	}
	c.CurrentWeight += weight
	c.Mu.Unlock()

	return fn()
}
